File: samples/scala-valueentity-shopping-cart/src/main/scala/com/example/shoppingcart/ShoppingCartActionImpl.scala
package com.example.shoppingcart

import com.akkaserverless.scalasdk.action.Action
import com.akkaserverless.scalasdk.action.ActionCreationContext
import com.google.protobuf.empty.Empty

import java.util.UUID
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** An action. */
class ShoppingCartActionImpl(creationContext: ActionCreationContext) extends AbstractShoppingCartAction {

  // tag::initialize[]
  override def initializeCart(newCart: NewCart): Action.Effect[NewCartCreated] = {
    val cartId = UUID.randomUUID().toString // <1>

    val created: Future[Empty] =
      components.shoppingCart.create(CreateCart(cartId)).execute() // <2>

    val effect: Future[Action.Effect[NewCartCreated]] = // <3>
      created.map(_ => effects.reply(NewCartCreated(cartId))) // <4>
        .recover(_ => effects.error("Failed to create cart, please retry")) // <5>

    effects.asyncEffect(effect) // <6>
  }
  // end::initialize[]
}

File: samples/java-valueentity-shopping-cart/src/main/java/com/example/shoppingcart/ShoppingCartActionImpl.java
package com.example.shoppingcart;

import com.akkaserverless.javasdk.action.ActionCreationContext;
import com.google.protobuf.Empty;

import java.util.UUID;
import java.util.concurrent.CompletionStage;

/** An action. */
public class ShoppingCartActionImpl extends AbstractShoppingCartAction {

  public ShoppingCartActionImpl(ActionCreationContext creationContext) {}

  // tag::initialize[]
  @Override
  public Effect<ShoppingCartController.NewCartCreated> initializeCart(ShoppingCartController.NewCart newCart) {
    final String id = UUID.randomUUID().toString(); // <1>
    CompletionStage<Empty> shoppingCartCreated =
        components().shoppingCart().create(ShoppingCartApi.CreateCart.newBuilder().setCartId(id).build()) // <2>
        .execute(); // <3>

    // transform response
    CompletionStage<Effect<ShoppingCartController.NewCartCreated>> effect =
        shoppingCartCreated.handle((empty, error) -> { // <4>
          if (error == null) {
            return effects().reply(ShoppingCartController.NewCartCreated.newBuilder().setCartId(id).build()); // <5>
          } else {
            return effects().error("Failed to create cart, please retry"); // <6>
          }
        });

    return effects().asyncEffect(effect); // <7>
  }
  // end::initialize[]
}
